AsyncCollectiveTensor: dont sync on view ops

bdhirsh opened the pull request:
AsyncCollectiveTensor is a tensor subclass that is meant to "delay synchronization" when you call into the functional collectives API's. It does this (if I understand correctly) by internally holding an "unsynchronized" version of the tensor, which is the result of the communication op, and internally calling `.wait()` to synchronize the data the next time it is used.

Previously, these wait() calls would happen immediately, because `AsyncCollectiveTensor` gets wrapped by `DTensor()`, which calls `.detach()` on its inner tensor, immediately causing the sync (code: https://github.com/pytorch/pytorch/blob/1518d5eec4425b74b6114d18dd8998df20d0fb0a/torch/distributed/_tensor/api.py#L207)

AsyncCollectiveTensor shouldn't need to do a synchronization if you try to detach() it though - in fact, it should be fine to avoid synchronizing if you perform any view ops on it (which just require viewing metadata, but not actual data). This PR tries to update `AsyncCollectiveTensor` to delay `wait()` calls whenever the subclass encounters a view op.

Added some light testing, that just runs some DTensor compute followed by view ops, and confirms that the output is still an `AsyncCollectiveTensor` when we call `.to_local()`.

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #105240

bdhirsh committed:
AsyncCollectiveTensor: dont sync on view ops

[ghstack-poisoned]

bdhirsh committed:
Update on "AsyncCollectiveTensor: dont sync on view ops"




[ghstack-poisoned]

bdhirsh committed:
Update on "AsyncCollectiveTensor: dont sync on view ops"

AsyncCollectiveTensor is a tensor subclass that is meant to "delay synchronization" when you call into the functional collectives API's. It does this (if I understand correctly) by internally holding an "unsynchronized" version of the tensor, which is the result of the communication op, and internally calling `.wait()` to synchronize the data the next time it is used.

Previously, these wait() calls would happen immediately, because `AsyncCollectiveTensor` gets wrapped by `DTensor()`, which calls `.detach()` on its inner tensor, immediately causing the sync (code: https://github.com/pytorch/pytorch/blob/1518d5eec4425b74b6114d18dd8998df20d0fb0a/torch/distributed/_tensor/api.py#L207)

AsyncCollectiveTensor shouldn't need to do a synchronization if you try to detach() it though - in fact, it should be fine to avoid synchronizing if you perform any view ops on it (which just require viewing metadata, but not actual data). This PR tries to update `AsyncCollectiveTensor` to delay `wait()` calls whenever the subclass encounters a view op.

Added some light testing, that just runs some DTensor compute followed by view ops, and confirms that the output is still an `AsyncCollectiveTensor` when we call `.to_local()`.




[ghstack-poisoned]

bdhirsh committed:
Update on "AsyncCollectiveTensor: dont sync on view ops"

AsyncCollectiveTensor is a tensor subclass that is meant to "delay synchronization" when you call into the functional collectives API's. It does this (if I understand correctly) by internally holding an "unsynchronized" version of the tensor, which is the result of the communication op, and internally calling `.wait()` to synchronize the data the next time it is used.

Previously, these wait() calls would happen immediately, because `AsyncCollectiveTensor` gets wrapped by `DTensor()`, which calls `.detach()` on its inner tensor, immediately causing the sync (code: https://github.com/pytorch/pytorch/blob/1518d5eec4425b74b6114d18dd8998df20d0fb0a/torch/distributed/_tensor/api.py#L207)

AsyncCollectiveTensor shouldn't need to do a synchronization if you try to detach() it though - in fact, it should be fine to avoid synchronizing if you perform any view ops on it (which just require viewing metadata, but not actual data). This PR tries to update `AsyncCollectiveTensor` to delay `wait()` calls whenever the subclass encounters a view op.

Added some light testing, that just runs some DTensor compute followed by view ops, and confirms that the output is still an `AsyncCollectiveTensor` when we call `.to_local()`.




[ghstack-poisoned]

bdhirsh committed:
Update on "AsyncCollectiveTensor: dont sync on view ops"

AsyncCollectiveTensor is a tensor subclass that is meant to "delay synchronization" when you call into the functional collectives API's. It does this (if I understand correctly) by internally holding an "unsynchronized" version of the tensor, which is the result of the communication op, and internally calling `.wait()` to synchronize the data the next time it is used.

Previously, these wait() calls would happen immediately, because `AsyncCollectiveTensor` gets wrapped by `DTensor()`, which calls `.detach()` on its inner tensor, immediately causing the sync (code: https://github.com/pytorch/pytorch/blob/1518d5eec4425b74b6114d18dd8998df20d0fb0a/torch/distributed/_tensor/api.py#L207)

AsyncCollectiveTensor shouldn't need to do a synchronization if you try to detach() it though - in fact, it should be fine to avoid synchronizing if you perform any view ops on it (which just require viewing metadata, but not actual data). This PR tries to update `AsyncCollectiveTensor` to delay `wait()` calls whenever the subclass encounters a view op.

Added some light testing, that just runs some DTensor compute followed by view ops, and confirms that the output is still an `AsyncCollectiveTensor` when we call `.to_local()`.




[ghstack-poisoned]

bdhirsh committed:
Update on "AsyncCollectiveTensor: dont sync on view ops"

AsyncCollectiveTensor is a tensor subclass that is meant to "delay synchronization" when you call into the functional collectives API's. It does this (if I understand correctly) by internally holding an "unsynchronized" version of the tensor, which is the result of the communication op, and internally calling `.wait()` to synchronize the data the next time it is used.

Previously, these wait() calls would happen immediately, because `AsyncCollectiveTensor` gets wrapped by `DTensor()`, which calls `.detach()` on its inner tensor, immediately causing the sync (code: https://github.com/pytorch/pytorch/blob/1518d5eec4425b74b6114d18dd8998df20d0fb0a/torch/distributed/_tensor/api.py#L207)

AsyncCollectiveTensor shouldn't need to do a synchronization if you try to detach() it though - in fact, it should be fine to avoid synchronizing if you perform any view ops on it (which just require viewing metadata, but not actual data). This PR tries to update `AsyncCollectiveTensor` to delay `wait()` calls whenever the subclass encounters a view op.

Added some light testing, that just runs some DTensor compute followed by view ops, and confirms that the output is still an `AsyncCollectiveTensor` when we call `.to_local()`.




[ghstack-poisoned]

wanchaol commented on test/distributed/_tensor/test_dtensor.py:
```diff
@@ -244,6 +246,42 @@ def test_to_local(self):
+            dt_out_redistribute = dt_out.redistribute(mesh, [Replicate()])
+            # Make sure we haven't synced yet
+            self.assertTrue(_tensor_needs_wait(dt_out_redistribute))
+            dt_out_redistribute_view = dt_out_redistribute_view(
```
I think this should be `dt_out_redistribute.view(...)` instead?

wanchaol reviewed (approved):
looks good, stamp to unblock. one nit might need fix in tests

bdhirsh committed:
Update on "AsyncCollectiveTensor: dont sync on view ops"

AsyncCollectiveTensor is a tensor subclass that is meant to "delay synchronization" when you call into the functional collectives API's. It does this (if I understand correctly) by internally holding an "unsynchronized" version of the tensor, which is the result of the communication op, and internally calling `.wait()` to synchronize the data the next time it is used.

Previously, these wait() calls would happen immediately, because `AsyncCollectiveTensor` gets wrapped by `DTensor()`, which calls `.detach()` on its inner tensor, immediately causing the sync (code: https://github.com/pytorch/pytorch/blob/1518d5eec4425b74b6114d18dd8998df20d0fb0a/torch/distributed/_tensor/api.py#L207)

AsyncCollectiveTensor shouldn't need to do a synchronization if you try to detach() it though - in fact, it should be fine to avoid synchronizing if you perform any view ops on it (which just require viewing metadata, but not actual data). This PR tries to update `AsyncCollectiveTensor` to delay `wait()` calls whenever the subclass encounters a view op.

Added some light testing, that just runs some DTensor compute followed by view ops, and confirms that the output is still an `AsyncCollectiveTensor` when we call `.to_local()`.




[ghstack-poisoned]

bdhirsh committed:
Update on "AsyncCollectiveTensor: dont sync on view ops"

AsyncCollectiveTensor is a tensor subclass that is meant to "delay synchronization" when you call into the functional collectives API's. It does this (if I understand correctly) by internally holding an "unsynchronized" version of the tensor, which is the result of the communication op, and internally calling `.wait()` to synchronize the data the next time it is used.

Previously, these wait() calls would happen immediately, because `AsyncCollectiveTensor` gets wrapped by `DTensor()`, which calls `.detach()` on its inner tensor, immediately causing the sync (code: https://github.com/pytorch/pytorch/blob/1518d5eec4425b74b6114d18dd8998df20d0fb0a/torch/distributed/_tensor/api.py#L207)

AsyncCollectiveTensor shouldn't need to do a synchronization if you try to detach() it though - in fact, it should be fine to avoid synchronizing if you perform any view ops on it (which just require viewing metadata, but not actual data). This PR tries to update `AsyncCollectiveTensor` to delay `wait()` calls whenever the subclass encounters a view op.

Added some light testing, that just runs some DTensor compute followed by view ops, and confirms that the output is still an `AsyncCollectiveTensor` when we call `.to_local()`.




[ghstack-poisoned]

bdhirsh committed:
Update on "AsyncCollectiveTensor: dont sync on view ops"

AsyncCollectiveTensor is a tensor subclass that is meant to "delay synchronization" when you call into the functional collectives API's. It does this (if I understand correctly) by internally holding an "unsynchronized" version of the tensor, which is the result of the communication op, and internally calling `.wait()` to synchronize the data the next time it is used.

Previously, these wait() calls would happen immediately, because `AsyncCollectiveTensor` gets wrapped by `DTensor()`, which calls `.detach()` on its inner tensor, immediately causing the sync (code: https://github.com/pytorch/pytorch/blob/1518d5eec4425b74b6114d18dd8998df20d0fb0a/torch/distributed/_tensor/api.py#L207)

AsyncCollectiveTensor shouldn't need to do a synchronization if you try to detach() it though - in fact, it should be fine to avoid synchronizing if you perform any view ops on it (which just require viewing metadata, but not actual data). This PR tries to update `AsyncCollectiveTensor` to delay `wait()` calls whenever the subclass encounters a view op.

Added some light testing, that just runs some DTensor compute followed by view ops, and confirms that the output is still an `AsyncCollectiveTensor` when we call `.to_local()`.




[ghstack-poisoned]

fduwjj commented on torch/distributed/_functional_collectives_impl.py:
```diff
@@ -87,6 +87,13 @@ def _wait_tensor(tensor: torch.Tensor) -> torch.Tensor:
     return tensor
 
+def _tensor_needs_wait(tensor: torch.Tensor) -> Optional[bool]:
+    data_ptr = tensor.data_ptr()
```
do you need: `global data_ptr_to_work` here?

kumpera replied:
no, you only need to use `global` if you assign to it.

fduwjj replied:
Interesting, because we use global in L83, too. That's why I am asking.

fduwjj commented on torch/distributed/_functional_collectives.py:
```diff
@@ -291,6 +291,15 @@ def reduce_scatter_tensor_coalesced(
+    if len(schema.arguments) > 0:
+        first_arg = schema.arguments[0]
+        # check if op is a view
+        return first_arg.alias_info is not None and not first_arg.alias_info.is_write
```
wow, til...

fduwjj reviewed (approved):
Nice, we will have the async collective for TP now. Thanks!!

bdhirsh committed:
Update on "AsyncCollectiveTensor: dont sync on view ops"

AsyncCollectiveTensor is a tensor subclass that is meant to "delay synchronization" when you call into the functional collectives API's. It does this (if I understand correctly) by internally holding an "unsynchronized" version of the tensor, which is the result of the communication op, and internally calling `.wait()` to synchronize the data the next time it is used.

Previously, these wait() calls would happen immediately, because `AsyncCollectiveTensor` gets wrapped by `DTensor()`, which calls `.detach()` on its inner tensor, immediately causing the sync (code: https://github.com/pytorch/pytorch/blob/1518d5eec4425b74b6114d18dd8998df20d0fb0a/torch/distributed/_tensor/api.py#L207)

AsyncCollectiveTensor shouldn't need to do a synchronization if you try to detach() it though - in fact, it should be fine to avoid synchronizing if you perform any view ops on it (which just require viewing metadata, but not actual data). This PR tries to update `AsyncCollectiveTensor` to delay `wait()` calls whenever the subclass encounters a view op.

Added some light testing, that just runs some DTensor compute followed by view ops, and confirms that the output is still an `AsyncCollectiveTensor` when we call `.to_local()`.




[ghstack-poisoned]

bdhirsh committed:
Update on "AsyncCollectiveTensor: dont sync on view ops"

AsyncCollectiveTensor is a tensor subclass that is meant to "delay synchronization" when you call into the functional collectives API's. It does this (if I understand correctly) by internally holding an "unsynchronized" version of the tensor, which is the result of the communication op, and internally calling `.wait()` to synchronize the data the next time it is used.

Previously, these wait() calls would happen immediately, because `AsyncCollectiveTensor` gets wrapped by `DTensor()`, which calls `.detach()` on its inner tensor, immediately causing the sync (code: https://github.com/pytorch/pytorch/blob/1518d5eec4425b74b6114d18dd8998df20d0fb0a/torch/distributed/_tensor/api.py#L207)

AsyncCollectiveTensor shouldn't need to do a synchronization if you try to detach() it though - in fact, it should be fine to avoid synchronizing if you perform any view ops on it (which just require viewing metadata, but not actual data). This PR tries to update `AsyncCollectiveTensor` to delay `wait()` calls whenever the subclass encounters a view op.

Added some light testing, that just runs some DTensor compute followed by view ops, and confirms that the output is still an `AsyncCollectiveTensor` when we call `.to_local()`.




[ghstack-poisoned]

bdhirsh committed:
Update on "AsyncCollectiveTensor: dont sync on view ops"

AsyncCollectiveTensor is a tensor subclass that is meant to "delay synchronization" when you call into the functional collectives API's. It does this (if I understand correctly) by internally holding an "unsynchronized" version of the tensor, which is the result of the communication op, and internally calling `.wait()` to synchronize the data the next time it is used.

Previously, these wait() calls would happen immediately, because `AsyncCollectiveTensor` gets wrapped by `DTensor()`, which calls `.detach()` on its inner tensor, immediately causing the sync (code: https://github.com/pytorch/pytorch/blob/1518d5eec4425b74b6114d18dd8998df20d0fb0a/torch/distributed/_tensor/api.py#L207)

AsyncCollectiveTensor shouldn't need to do a synchronization if you try to detach() it though - in fact, it should be fine to avoid synchronizing if you perform any view ops on it (which just require viewing metadata, but not actual data). This PR tries to update `AsyncCollectiveTensor` to delay `wait()` calls whenever the subclass encounters a view op.

Added some light testing, that just runs some DTensor compute followed by view ops, and confirms that the output is still an `AsyncCollectiveTensor` when we call `.to_local()`.




[ghstack-poisoned]

bdhirsh committed:
Update on "AsyncCollectiveTensor: dont sync on view ops"

AsyncCollectiveTensor is a tensor subclass that is meant to "delay synchronization" when you call into the functional collectives API's. It does this (if I understand correctly) by internally holding an "unsynchronized" version of the tensor, which is the result of the communication op, and internally calling `.wait()` to synchronize the data the next time it is used.

Previously, these wait() calls would happen immediately, because `AsyncCollectiveTensor` gets wrapped by `DTensor()`, which calls `.detach()` on its inner tensor, immediately causing the sync (code: https://github.com/pytorch/pytorch/blob/1518d5eec4425b74b6114d18dd8998df20d0fb0a/torch/distributed/_tensor/api.py#L207)

AsyncCollectiveTensor shouldn't need to do a synchronization if you try to detach() it though - in fact, it should be fine to avoid synchronizing if you perform any view ops on it (which just require viewing metadata, but not actual data). This PR tries to update `AsyncCollectiveTensor` to delay `wait()` calls whenever the subclass encounters a view op.

Added some light testing, that just runs some DTensor compute followed by view ops, and confirms that the output is still an `AsyncCollectiveTensor` when we call `.to_local()`.




[ghstack-poisoned]

bdhirsh committed:
Update on "AsyncCollectiveTensor: dont sync on view ops"

AsyncCollectiveTensor is a tensor subclass that is meant to "delay synchronization" when you call into the functional collectives API's. It does this (if I understand correctly) by internally holding an "unsynchronized" version of the tensor, which is the result of the communication op, and internally calling `.wait()` to synchronize the data the next time it is used.

Previously, these wait() calls would happen immediately, because `AsyncCollectiveTensor` gets wrapped by `DTensor()`, which calls `.detach()` on its inner tensor, immediately causing the sync (code: https://github.com/pytorch/pytorch/blob/1518d5eec4425b74b6114d18dd8998df20d0fb0a/torch/distributed/_tensor/api.py#L207)

AsyncCollectiveTensor shouldn't need to do a synchronization if you try to detach() it though - in fact, it should be fine to avoid synchronizing if you perform any view ops on it (which just require viewing metadata, but not actual data). This PR tries to update `AsyncCollectiveTensor` to delay `wait()` calls whenever the subclass encounters a view op.

Added some light testing, that just runs some DTensor compute followed by view ops, and confirms that the output is still an `AsyncCollectiveTensor` when we call `.to_local()`.




[ghstack-poisoned]

bdhirsh committed:
Update on "AsyncCollectiveTensor: dont sync on view ops"

AsyncCollectiveTensor is a tensor subclass that is meant to "delay synchronization" when you call into the functional collectives API's. It does this (if I understand correctly) by internally holding an "unsynchronized" version of the tensor, which is the result of the communication op, and internally calling `.wait()` to synchronize the data the next time it is used.

Previously, these wait() calls would happen immediately, because `AsyncCollectiveTensor` gets wrapped by `DTensor()`, which calls `.detach()` on its inner tensor, immediately causing the sync (code: https://github.com/pytorch/pytorch/blob/1518d5eec4425b74b6114d18dd8998df20d0fb0a/torch/distributed/_tensor/api.py#L207)

AsyncCollectiveTensor shouldn't need to do a synchronization if you try to detach() it though - in fact, it should be fine to avoid synchronizing if you perform any view ops on it (which just require viewing metadata, but not actual data). This PR tries to update `AsyncCollectiveTensor` to delay `wait()` calls whenever the subclass encounters a view op.

Added some light testing, that just runs some DTensor compute followed by view ops, and confirms that the output is still an `AsyncCollectiveTensor` when we call `.to_local()`.




[ghstack-poisoned]

wconstab commented on test/distributed/_tensor/test_dtensor.py:
```diff
@@ -245,6 +246,43 @@ def test_to_local(self):
+        y = torch.arange(8, requires_grad=True, dtype=torch.float32)
+        out = fn(x, y)
+        # Make sure we haven't synced yet
+        # self.assertTrue(_tensor_needs_wait(out))
```
should these be uncommented?

wconstab reviewed (approved):
LGTM, thanks for this fix!

fduwjj commented:
When I patched this PR to our 2D (DTensor based TP + FSDP), I got this error:
```
RuntimeError: The tensor has a non-zero number of elements, but its data is not allocated yet. Caffe2 uses a lazy allocation, so you will need to call mutable_data() or raw_mutable_data() to actually allocate memory.
```

Not sure if I am missing anything here.

bdhirsh commented:
@fduwjj there's definitely a bug in this PR somewhere. I haven't had a chance to try to fix it, because distributed was segfaulting locally for me when I tried to run the test. Unfortunately I won't have any time the rest of this week to take more of a look.

My first guess is that: whenever we delay a view, we probably need to stash it somewhere so we remember to lazily replay it once the tensor is synced (right now my PR is not doing that).

wanchaol commented:
@fduwjj I'm helping @bdhirsh debugging the issues, I'll send out a update soon

wanchaol committed:
Update on "AsyncCollectiveTensor: dont sync on view ops"

AsyncCollectiveTensor is a tensor subclass that is meant to "delay synchronization" when you call into the functional collectives API's. It does this (if I understand correctly) by internally holding an "unsynchronized" version of the tensor, which is the result of the communication op, and internally calling `.wait()` to synchronize the data the next time it is used.

Previously, these wait() calls would happen immediately, because `AsyncCollectiveTensor` gets wrapped by `DTensor()`, which calls `.detach()` on its inner tensor, immediately causing the sync (code: https://github.com/pytorch/pytorch/blob/1518d5eec4425b74b6114d18dd8998df20d0fb0a/torch/distributed/_tensor/api.py#L207)

AsyncCollectiveTensor shouldn't need to do a synchronization if you try to detach() it though - in fact, it should be fine to avoid synchronizing if you perform any view ops on it (which just require viewing metadata, but not actual data). This PR tries to update `AsyncCollectiveTensor` to delay `wait()` calls whenever the subclass encounters a view op.

Added some light testing, that just runs some DTensor compute followed by view ops, and confirms that the output is still an `AsyncCollectiveTensor` when we call `.to_local()`.




[ghstack-poisoned]

wanchaol committed:
Update on "AsyncCollectiveTensor: dont sync on view ops"

AsyncCollectiveTensor is a tensor subclass that is meant to "delay synchronization" when you call into the functional collectives API's. It does this (if I understand correctly) by internally holding an "unsynchronized" version of the tensor, which is the result of the communication op, and internally calling `.wait()` to synchronize the data the next time it is used.

Previously, these wait() calls would happen immediately, because `AsyncCollectiveTensor` gets wrapped by `DTensor()`, which calls `.detach()` on its inner tensor, immediately causing the sync (code: https://github.com/pytorch/pytorch/blob/1518d5eec4425b74b6114d18dd8998df20d0fb0a/torch/distributed/_tensor/api.py#L207)

AsyncCollectiveTensor shouldn't need to do a synchronization if you try to detach() it though - in fact, it should be fine to avoid synchronizing if you perform any view ops on it (which just require viewing metadata, but not actual data). This PR tries to update `AsyncCollectiveTensor` to delay `wait()` calls whenever the subclass encounters a view op.

Added some light testing, that just runs some DTensor compute followed by view ops, and confirms that the output is still an `AsyncCollectiveTensor` when we call `.to_local()`.




[ghstack-poisoned]

wanchaol committed:
Update on "AsyncCollectiveTensor: dont sync on view ops"

AsyncCollectiveTensor is a tensor subclass that is meant to "delay synchronization" when you call into the functional collectives API's. It does this (if I understand correctly) by internally holding an "unsynchronized" version of the tensor, which is the result of the communication op, and internally calling `.wait()` to synchronize the data the next time it is used.

Previously, these wait() calls would happen immediately, because `AsyncCollectiveTensor` gets wrapped by `DTensor()`, which calls `.detach()` on its inner tensor, immediately causing the sync (code: https://github.com/pytorch/pytorch/blob/1518d5eec4425b74b6114d18dd8998df20d0fb0a/torch/distributed/_tensor/api.py#L207)

AsyncCollectiveTensor shouldn't need to do a synchronization if you try to detach() it though - in fact, it should be fine to avoid synchronizing if you perform any view ops on it (which just require viewing metadata, but not actual data). This PR tries to update `AsyncCollectiveTensor` to delay `wait()` calls whenever the subclass encounters a view op.

Added some light testing, that just runs some DTensor compute followed by view ops, and confirms that the output is still an `AsyncCollectiveTensor` when we call `.to_local()`.




[ghstack-poisoned]

wanchaol commented:
@pytorchbot merge -f "failures are unrelated to this PR"

bdhirsh commented:
Thanks for the fix @wanchaol 😄
